cli/azd/pkg/auth/manager.go

```diff
@@ -0,0 +1,433 @@
+		return nil, fmt.Errorf("creating msal client: %w", err)
+	}
+
+	return &Manager{
+		publicClient:    &msalPublicClientAdapter{client: &publicClientApp},
+		configManager:   configManager,
+		credentialCache: newCredentialCache(authRoot),
+	}, nil
+}
+
+var ErrNoCurrentUser = errors.New("not logged in, run `azd login` to login")
+
+// EnsureLoggedInCredential uses the credential's GetToken method to ensure an access token can be fetched. If this fails,
+// nil, ErrNoCurrentUser is returned. On success, the token we fetched is returned.
+func EnsureLoggedInCredential(ctx context.Context, credential azcore.TokenCredential) (*azcore.AccessToken, error) {
+	token, err := credential.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: cLoginScopes,
+	})
+	if err != nil {
+		return &azcore.AccessToken{}, ErrNoCurrentUser
+	}
+
+	return &token, nil
+}
+
+// CredentialForCurrentUser returns a TokenCredential instance for the current user. If `auth.useLegacyAzCliAuth` is set to
+// a truthy value in config, an instance of azidentity.AzureCLICredential is returned instead.
+func (m *Manager) CredentialForCurrentUser(ctx context.Context) (azcore.TokenCredential, error) {
+	cfg, err := config.GetUserConfig(m.configManager)
+	if err != nil {
+		return nil, fmt.Errorf("fetching current user: %w", err)
+	}
+
+	if useLegacyAuth, has := cfg.Get(cUseLegacyAzCliAuthKey); has {
+		if use, err := strconv.ParseBool(useLegacyAuth.(string)); err == nil && use {
+			log.Printf("delegating auth to az since %s is set to true", cUseLegacyAzCliAuthKey)
+			cred, err := azidentity.NewAzureCLICredential(nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create credential: %v: %w", err, ErrNoCurrentUser)
+			}
+			return cred, nil
+		}
+	}
+
+	currentUser, err := readUserProperties(cfg)
+	if err != nil {
+		return nil, ErrNoCurrentUser
+	}
+
+	if currentUser.HomeAccountID != nil {
+		for _, account := range m.publicClient.Accounts() {
+			if account.HomeAccountID == *currentUser.HomeAccountID {
+				return newAzdCredential(m.publicClient, &account), nil
+			}
+		}
+	} else if currentUser.TenantID != nil && currentUser.ClientID != nil {
+		ps, err := m.loadSecret(*currentUser.TenantID, *currentUser.ClientID)
+		if err != nil {
+			return nil, fmt.Errorf("loading secret: %v: %w", err, ErrNoCurrentUser)
+		}
+
+		if ps.ClientSecret != nil {
+			cred, err := azidentity.NewClientSecretCredential(
+				*currentUser.TenantID, *currentUser.ClientID, *ps.ClientSecret, nil,
+			)
+
+			if err != nil {
+				return nil, fmt.Errorf("creating credential: %v: %w", err, ErrNoCurrentUser)
+			}
+
+			return cred, nil
+
+		} else if ps.ClientCertificate != nil {
+			certData, err := base64.StdEncoding.DecodeString(*ps.ClientCertificate)
+			if err != nil {
+				return nil, fmt.Errorf("decoding certificate: %w", err)
+			}
+
+			certs, key, err := azidentity.ParseCertificates(certData, nil)
+			if err != nil {
+				return nil, fmt.Errorf("parsing certificate: %w", err)
+			}
```

Comment:
Would this be fixed with some logic action? Should probably provide some sort of user guidance here

Reply:
I changed these returns to look like this:

```
fmt.Errorf("parsing certificate: %v: %w", err, ErrNoCurrentUser)
```

So that now `ErrNoCurrentUser` is wrapped and folks will see an actionable message telling them they need to `azd login`.  This error path would only happen in the value we stored in the credential cache becomes corrupted somehow (we store it after validating it in the login flow).  The fix is to log in again which will require you to present valid authentication data which we will then store, overwriting our corrupted copy.

Reply:
I see. I think the part that may be confusing is that we would surface `ErrNoCurrentUser` in the login command itself, which would end up saying "logging in: not logged in, run `azd login` to login".

I wonder if we need to wrap the `ErrNoCurrentUser` errors in the `login` command to be slightly different.